README.md
# Magnetic Button

The Magnetic Button is a custom Flutter widget that creates an interactive button with a "magnetic" effect. When you hover over the button, it moves in response to the mouse position, creating a unique and engaging user experience.

#### Platform Support

| Android | iOS | MacOS | Web | Linux | Windows |
| :-----: | :-: | :---: | :-: | :---: | :-----: |
|   ❓   | ❓  |  ✅   | ✅ |  ✅   |  ✅    |


![preview](https://github.com/darksabotage1/assets/assets/136040945/f1bb1ebf-2d10-4f20-badb-9f00e703e625)


## Untested Features

- **Mobile On-Hold Magnetic Effect(Undergoing Testing)**: We are currently in the testing phase for an on-hold magnetic effect specifically designed for mobile devices. This functionality is intended to enable buttons to maintain their movement in response to hold gestures, enhancing interactivity on mobile platforms."

## Features

- **Hover Effect**: The button moves in response to the mouse position when hovered over, creating a "magnetic" effect.
- **Return Duration**: When the mouse is not hovering over the widget, the button returns to its original position. The speed at which it returns is controlled by the `duration` property. This property is set in milliseconds, so a larger value will make the return slower, and a smaller value will make it faster.
- **Customizable**: You can customize the child widget of the Magnetic Button to fit your needs.
- **Max Movement**: You can control the maximum movement of the button in response to the mouse position using the mx and my properties.
- **Distance to Trigger**: A customizable parameter that determines the proximity at which an animation for a widget is initiated.

## Usage

1. Create a `MagneticButton` widget and pass in your desired child widget:

   ```dart
   Offset values = const Offset(0, 0);
   MagneticButton(
          duration: const Duration(milliseconds: 100),
          onChanged: (Offset value) {
            setState(() => values = value);
          },
          child: Container(
            decoration: BoxDecoration(
                borderRadius: const BorderRadius.all(Radius.circular(40)),
                color: Colors.white,
                border: Border.all(color: Colors.black)),
            width: 160,
            height: 80,
            child: Center(
              child: Transform.translate(
                offset: Offset(values.dx / 4, values.dy / 4),
                child: const Text(
                  'Explore',
                  style: TextStyle(
                    fontWeight: FontWeight.w400,
                    fontSize: 19,
                  ),
                ),
              ),
            ),
          ),
        ),

## Future Features

We have successfully integrated all the proposed features for the Magnetic Button, except ‘Mobile On-Hold Magnetic Effect’ feature, which is currently undergoing testing. At this point, we have no additional features planned for future implementation. However, we welcome feedback and suggestions from our users. If you have an idea for a feature that you’d like to see in Magnetic Button, please let us know! 🙂. 

CHANGELOG.md
## 0.0.1

* Magnetic Button initial release.

## 0.0.2

* Description Update

## 0.0.3

* Initial Feature Updates with Bugs

## 0.0.4
### Added
- New properties for the magnetic widget:
  - `duration`: The duration of the animation.
  - `curve`: The curve of the animation.
  - `height`: The height of the magnetic widget. This is an optional parameter.
  - `width`: The width of the magnetic widget. This is an optional parameter.
  - `padding`: The padding around the magnetic widget. This is an optional parameter.
  - `mobile`: A boolean value indicating whether the widget should respond to long press events on mobile(not yet tested for functionality). If false, only Web will work.
### Improved
- Performance optimizations in the code:
  - reduced the number of times renderBox.localToGlobal(Offset.zero) is called
  - moved the calculations for relX and relY inside the if (distanceMouseButton < distanceToTrigger) statement, so they’re only calculated if necessary.

## 0.0.5
### Added
- New properties for the magnetic widget:
  - `onChanged`: A nullable callback function called when an `Offset` change event occurs.

## 0.0.6
### Added
* Major Bug Fix

## 0.0.7
### Added
* Bug Fix & No need to use Global Key 
- New properties for the magnetic widget:
  - `mx`: This is the horizontal offset of the magnetic widget
  - `my`: This is the vertical offset of the magnetic widget

## 0.0.8
### Added
* Minor Bug Fix
- New property for the magnetic widget::
  - `distance`: Determines the proximity at which an animation for a widget is initiated.
 

### To Do
- test hold feature on a live mobile device.


test/magnetic_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:magnetic_button/magnetic_button.dart';

void main() {
  testWidgets('Check MagneticWidget exists', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Center(
            child: MagneticButton(
              onChanged: (Offset value) {},
              distance: 0.8,
              child: Container(
                decoration: BoxDecoration(
                    borderRadius: const BorderRadius.all(Radius.circular(40)),
                    color: Colors.white,
                    border: Border.all(color: const Color(0xFF303032))),
                width: 160,
                height: 80,
                child: const Center(
                  child: Text(
                    'Explore',
                    style: TextStyle(
                      fontWeight: FontWeight.w400,
                      fontSize: 19,
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );

    final magneticFinder = find.byType(MagneticButton);
    expect(magneticFinder, findsOneWidget);
  });
}


lib/magnetic_button.dart
library magnetic_button;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'components/utils.dart';
import 'package:flutter/foundation.dart' show kIsWeb;

class MagneticButton extends StatefulWidget {
  final Widget child;

  /// This is the horizontal offset of the magnetic widget. It determines how much the button moves in the horizontal direction (left or right) when it’s being interacted with. A larger value will result in a greater horizontal movement.
  final double mx;

  /// This is the vertical offset of the magnetic widget. It controls how much the button moves in the vertical direction (up or down) during interaction. A larger value will result in a greater vertical movement
  final double my;

  /// The duration of the animation.
  final Duration duration;

  /// Determines the proximity at which an animation for a widget is initiated. Default value is 70% of the width of the render box.
  final double distance;

  /// The curve of the animation.
  final Curve curve;

  /// The height of the magnetic widget. This is an optional parameter.
  final double? height;

  /// The width of the magnetic widget. This is an optional parameter.
  final double? width;

  /// The padding around the magnetic widget. This is an optional parameter.
  final EdgeInsets? padding;

  /// A boolean value indicating whether the widget should respond to long press events on mobile. If false only Web will work.
  final bool mobile;

  /// A nullable callback function called when an `Offset` change event occurs.
  final ValueChanged<Offset>? onChanged;

  const MagneticButton({
    Key? key,
    required this.child,
    this.mx = 0.2,
    this.my = 0.2,
    this.duration = const Duration(milliseconds: 200),
    this.curve = Curves.easeOutCirc,
    this.height,
    this.width,
    this.padding,
    this.mobile = true,
    this.onChanged,
    this.distance = 0.7,
  }) : super(key: key);

  @override
  State<MagneticButton> createState() => MagneticButtonState();
}

class MagneticButtonState extends State<MagneticButton>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _animation;
  final Key key = GlobalKey();

  double _textX = 0.0;
  double _textY = 0.0;

  bool mouseIsHovering = false;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 600));
    // Use an easing curve for the animation.
    _animation = CurvedAnimation(
      parent: _animationController,
      curve: widget.curve,
      reverseCurve: widget.curve,
    );

    _animationController.addListener(() {
      setState(() {});
    });
    _animationController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        Future.delayed(widget.duration, () {
          if (_animationController.status != AnimationStatus.forward) {
            _animationController.reverse();
          }
        });
      }
    });

    WidgetsBinding.instance.addPostFrameCallback((_) {
      _startMouseListeners();
    });
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _handleMouseEnter(PointerEnterEvent event) {}

  void _handleMouseLeave(PointerExitEvent event) {
    mouseIsHovering = false;
    // Check if mouse is still hovering over inner MagneticButton
    if (!mouseIsHovering) {
      setState(() {
        _textX = 0.0;
        _textY = 0.0;
        if (widget.onChanged != null) widget.onChanged!(const Offset(0, 0));
      });
      if (_animationController.status != AnimationStatus.forward) {
        _animationController.forward();
      }
    }
  }

  void _startMouseListeners() {
    // Listen to mouse events.
    RendererBinding.instance.pointerRouter.addGlobalRoute((PointerEvent event) {
      if (event is PointerHoverEvent) {
        _handleMouseMove(event);
      } else if (event is PointerEnterEvent) {
        _handleMouseEnter(event);
      } else if (event is PointerExitEvent) {
        _handleMouseLeave(event);
      }
    });
  }

  void _handleMouseMove(PointerHoverEvent event) {
    if (!mounted) return;

    // Get the RenderBox of the widget.
    final RenderBox renderBox = context.findRenderObject() as RenderBox;

    // Calculate the distance to trigger the animation.
    final double distanceToTrigger = renderBox.size.width * widget.distance;

    // Get the global position of the widget.
    final Offset globalPosition = renderBox.localToGlobal(Offset.zero);

    final double distanceMouseButton = distance(
        event.position.dx,
        event.position.dy,
        globalPosition.dx + renderBox.size.width / 2,
        globalPosition.dy + renderBox.size.height / 2);

    if (distanceMouseButton < distanceToTrigger) {
      // Calculate the position of the mouse relative to the center of the button.
      final double relX =
          event.position.dx - (globalPosition.dx + renderBox.size.width / 2);
      final double relY =
          event.position.dy - (globalPosition.dy + renderBox.size.height / 2);

      // If the mouse is close enough, move the button and its text.
      setState(() {
        _textX = relX * widget.mx;
        _textY = relY * widget.my;
        if (widget.onChanged != null) widget.onChanged!(Offset(_textX, _textY));
      });
    } else if (_textX != 0.0 || _textY != 0.0) {
      // If the mouse is not close enough, reset the button and its text to their original positions.
      setState(() {
        _textX = 0.0;
        _textY = 0.0;
        if (widget.onChanged != null) widget.onChanged!(Offset(_textX, _textY));
      });
    }
  }

  void _handleHoldMove(LongPressMoveUpdateDetails details) {
    if (!mounted) return;

    // Get the RenderBox of the widget.
    final RenderBox renderBox = context.findRenderObject() as RenderBox;

    // Calculate the distance to trigger the animation.
    final double distanceToTrigger = renderBox.size.width * widget.distance;

    // Get the global position of the widget.
    final Offset globalPosition = renderBox.localToGlobal(Offset.zero);

    final double distanceMouseButton = distance(
        details.localPosition.dx,
        details.localPosition.dy,
        globalPosition.dx + renderBox.size.width / 2,
        globalPosition.dy + renderBox.size.height / 2);

    if (distanceMouseButton < distanceToTrigger) {
      // Calculate the position of the hold position relative to the center of the button.
      final double relX = details.localPosition.dx -
          (globalPosition.dx + renderBox.size.width / 2);
      final double relY = details.localPosition.dy -
          (globalPosition.dy + renderBox.size.height / 2);

      // If the hold position is close enough, move the button and its text.
      setState(() {
        _textX = relX * widget.mx;
        _textY = relY * widget.my;
        if (widget.onChanged != null) widget.onChanged!(Offset(_textX, _textY));
      });
    } else if (_textX != 0.0 || _textY != 0.0) {
      // If the hold position is not close enough, reset the button and its text to their original positions.
      setState(() {
        _textX = 0.0;
        _textY = 0.0;
        if (widget.onChanged != null) widget.onChanged!(Offset(_textX, _textY));
      });
    }
  }

  // If hold is not pressed or exit, reset the button and its text to their original positions.
  void _handleHoldLeave() {
    setState(() {
      _textX = 0.0;
      _textY = 0.0;
      widget.onChanged!(Offset(_textX, _textY));
    });

    if (_animationController.status != AnimationStatus.forward) {
      _animationController.forward();
    }
  }

  @override
  Widget build(BuildContext context) {
    return kIsWeb
        ? _buildWeb()
        : widget.mobile
            ? _buildMobile()
            : _buildWeb();
  }

  Widget _buildWeb() {
    return MouseRegion(
        onEnter: _handleMouseEnter,
        onExit: _handleMouseLeave,
        onHover: _handleMouseMove,
        child: _buildAnimatedContainer());
  }

  Widget _buildMobile() {
    return GestureDetector(
      onLongPressMoveUpdate: (details) => _handleHoldMove(details),
      onLongPressEnd: (details) => _handleHoldLeave(),
      child: _buildAnimatedContainer(),
    );
  }

  Widget _buildAnimatedContainer() {
    return AnimatedContainer(
      key: key,
      duration: widget.duration,
      height: widget.height,
      width: widget.width,
      padding: widget.padding,
      transform: Matrix4.translationValues(
        _textX * (1 - _animation.value),
        _textY * (1 - _animation.value),
        0.0,
      ),
      child: widget.child,
    );
  }
}


lib/components/use_mouse_position.dart
import 'package:flutter/material.dart';

class MousePositionResult {
  late final ValueNotifier<double?> mouseX;
  late final ValueNotifier<double?> mouseY;

  MousePositionResult({required this.mouseX, required this.mouseY});

  void updateMousePosition(PointerEvent event) {
    mouseX.value = event.position.dx;
    mouseY.value = event.position.dy;
  }
}

MousePositionResult useMousePosition() {
  final mouseX = ValueNotifier<double?>(null);
  final mouseY = ValueNotifier<double?>(null);

  final result = MousePositionResult(
    mouseX: mouseX,
    mouseY: mouseY,
  );

  return result;
}

class MousePositionExample extends StatelessWidget {
  final Widget child;

  const MousePositionExample({
    Key? key,
    required this.child,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final mousePositionResult = useMousePosition();

    return MouseRegion(
      onEnter: (event) => mousePositionResult.updateMousePosition(event),
      onExit: (event) => mousePositionResult.updateMousePosition(event),
      onHover: (event) => mousePositionResult.updateMousePosition(event),
      child: child,
    );
  }
}

// class MousePositionWidget extends StatefulWidget {
//   final Widget child;

//   const MousePositionWidget({
//     Key? key,
//     required this.child,
//   }) : super(key: key);
//   @override
//   State<MousePositionWidget> createState() => _MousePositionWidgetState();
// }

// class _MousePositionWidgetState extends State<MousePositionWidget> {
//   @override
//   Widget build(BuildContext context) {
//     Offset _lastPointerMoveLocation;
//     return Listener(
//       onPointerMove: (PointerMoveEvent event) {
//         setState(() {
//           _lastPointerMoveLocation = event.position;
//         });
//       },
//       child: widget.child,
//     );
//   }
// }


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:magnetic_button/magnetic_button.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Magnetic Button',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => HomePageState();
}

class HomePageState extends State<HomePage> {
  Offset values = const Offset(0, 0);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: MagneticButton(
          duration: const Duration(milliseconds: 100),
          onChanged: (Offset value) {
            setState(() => values = value);
          },
          distance: 0.8,
          child: Container(
            decoration: BoxDecoration(
                borderRadius: const BorderRadius.all(Radius.circular(40)),
                color: Colors.white,
                border: Border.all(color: Colors.black)),
            width: 160,
            height: 80,
            child: Center(
              child: Transform.translate(
                offset: Offset(values.dx / 4, values.dy / 4),
                child: const Text(
                  'Explore',
                  style: TextStyle(
                    fontWeight: FontWeight.w400,
                    fontSize: 19,
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
